### PR TITLE
MusicApp api

### PR DESCRIPTION
**Pull request by cp-priyanka-g**
Jwt authentication and other api

### CODE REVIEW

File: album/album.go [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+	UpdatedAt   string `json:"updated_at "`
+	ArtistId    int    `json:"artist_id"`
+}
+
+type AlbumTrack struct {
+	Id      int   `json:"id"`
+	AlbumId int   `json:"album_id"`
+	TrackId []int `json:"track_id"`
+}
+type AlbumTrackRemove struct {
```
Comment (cp-sumi-k) [inline album/album.go]: Why do we need this structure? For removing track, can't we take trackId's array?

File: album/album.go [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+	if err != nil {
+		fmt.Println("Message :", err)
+		return
+	}
+
+	id, err := rs.LastInsertId()
+	if err != nil {
+		log.Fatalln(err)
+	}
+	fmt.Println("id is :", id)
```
Comment (cp-sumi-k) [inline album/album.go]: We do not need to print log here. Remove it

File: album/album.go [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+
+	rows, err := rs.RowsAffected()
+	if err != nil {
+		log.Printf("Error %s when finding rows affected", err)
+		return
+	}
+	fmt.Println("Rows Affected:", rows)
```
Comment (cp-sumi-k) [inline album/album.go]: Remove extra logging

File: main.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+	{
+		apiRoutes.POST("/v1/artist", artistRepo.Create)
+		apiRoutes.PUT("/v1/artist/:id", artistRepo.Update)
+
+		apiRoutes.GET("/v1/artist/show", artistRepo.Read)
+		apiRoutes.DELETE("/v1/artist/remove", artistRepo.Delete)
+	}
+
+	//ALBUM
+	router.POST("/api/v1/album", albumRepo.Create)
```
Comment (cp-sumi-k) [inline main.go]: Don't we need to use authorization in these following APIs?

File: main.go [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+	router := gin.Default()
+	//router.Use(AuthorizeJWT)
+
+	var loginService register.LoginService = register.StaticLoginService()
+	var jwtService register.JWTService = register.JWTAuthService()
+	var loginController register.LoginController = register.LoginHandler(loginService, jwtService)
+
+	router.POST("/api/login", func(ctx *gin.Context) {
+		token := loginController.Login(ctx)
+		if token != "" {
```
Comment (cp-sumi-k) [inline main.go]: Why do we need this condition?

File: main.go [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+	}
+}
+func setupRouter(sqlDb *sqlx.DB) *gin.Engine {
+	router := gin.Default()
+	//router.Use(AuthorizeJWT)
+
+	var loginService register.LoginService = register.StaticLoginService()
+	var jwtService register.JWTService = register.JWTAuthService()
+	var loginController register.LoginController = register.LoginHandler(loginService, jwtService)
```
Comment (cp-sumi-k) [inline main.go]: Why do we need to use a login handler in this way? I think it is not proper way to work with repositories

File: main.go [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+	router.POST("/api/v1/register/admin-register", registerRepo.RegisterAdmin)
+	//router.POST("/api/v1/login", registerRepo.Login)
+
+	//ARTISTAdminauthorized
+	apiRoutes := router.Group("/api", AuthorizeJWT())
+	{
+		apiRoutes.POST("/v1/artist", artistRepo.Create)
+		apiRoutes.PUT("/v1/artist/:id", artistRepo.Update)
+
+		apiRoutes.GET("/v1/artist/show", artistRepo.Read)
```
Comment (cp-sumi-k) [inline main.go]: We can rename it like  `/v1/artists`

File: main.go [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+		apiRoutes.DELETE("/v1/artist/remove", artistRepo.Delete)
```
Comment (cp-sumi-k) [inline main.go]: No need to add `remove` here as we are using `delete` method already

File: main.go [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+	}
+
+	//ALBUM
+	router.POST("/api/v1/album", albumRepo.Create)
+	router.PUT("/api/v1/album/edit", albumRepo.Update)
+	router.DELETE("/api/v1/album/remove", albumRepo.Delete)
+	router.GET("/api/v1/album/show", albumRepo.Read)
+	router.POST("/api/v1/album/add", albumRepo.Add)
```
Comment (cp-sumi-k) [inline main.go]: What is the difference between `add` and `create` methods? Is it serving the same purpose?

File: album/album.go [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+	if err != nil {
+		panic(err)
+	}
+
+	c.JSON(http.StatusOK, gin.H{"Message": "Album  Removed Successfully"})
+}
+
+// Adding multiple coulmn /remove Track from Album
+
+func (repository *AlbumRepository) Add(c *gin.Context) {
```
Comment (cp-sumi-k) [inline album/album.go]: Make appropriate method name according to its functionality

File: album/album.go [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+
+	rows, err := rs.RowsAffected()
+	if err != nil {
+		log.Printf("Error %s when finding rows affected", err)
+		return
+	}
+	fmt.Println("Rows Affected:", rows)
+
+	defer stmt.Close()
+	return
```
Comment (cp-sumi-k) [inline album/album.go]: Isn't it the gin handler function that we called from API?
All API calls need some response in return.
So We should return JSON response from here not only `return`

File: album/album.go [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+
+	err := c.ShouldBindWith(&input, binding.JSON)
+
+	if err != nil {
+		c.Abort()
+		c.JSON(http.StatusBadRequest, gin.H{"Message": err.Error()})
+		return
+	}
+
+	// _, err = repository.Db.Exec(`DELETE From AlbumTrack WHERE album_id=? and track_id=? `, input.AlbumId, input.TrackId)
```
Comment (cp-sumi-k) [inline album/album.go]: Remove commented code always

File: album/album.go [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+		log.Fatalln(err)
+	}
+	fmt.Println("id is :", id)
+
+	rows, err := rs.RowsAffected()
+	if err != nil {
+		log.Printf("Error %s when finding rows affected", err)
+		return
+	}
+	fmt.Println("Rows Affected:", rows)
```
Comment (cp-sumi-k) [inline album/album.go]: Remove extra logs

File: album/album.go [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+
+	defer stmt.Close()
+	return
+
+	//c.JSON(http.StatusOK, gin.H{"Message": "Track  Removed from album  Successfully"})
```
Comment (cp-sumi-k) [inline album/album.go]: This is valid response for  API, not only `return`

File: artist/artist.go [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+
+	err := c.ShouldBindWith(&input, binding.JSON)
+
+	if err != nil {
+		c.Abort()
+		c.JSON(http.StatusBadRequest, gin.H{"Message": err.Error()})
+		return
+	}
+
+	_, err = repository.Db.Exec(`INSERT INTO Artist(name,image_url) VALUES (?,?)`, input.Name, input.Image_url)
```
Comment (cp-sumi-k) [inline artist/artist.go:43]: What about `created_at` and `updated_at`?

File: artist/artist.go [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+
+	err := c.ShouldBindWith(&input, binding.JSON)
+
+	if err != nil {
+		c.Abort()
+		c.JSON(http.StatusBadRequest, gin.H{"Message": err.Error()})
+		return
+	}
+
+	_, err = repository.Db.Exec(`UPDATE Artist SET name=? ,image_url=? WHERE id=?`, input.Name, input.Image_url, input.Id)
```
Comment (cp-sumi-k) [inline artist/artist.go:88]: Shouldn't we have update `updated_at` here also?

File: favourite/favourite.go [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+
+	"github.com/gin-gonic/gin"
+	"github.com/gin-gonic/gin/binding"
+	_ "github.com/go-sql-driver/mysql"
+	"github.com/jmoiron/sqlx"
+)
+
+type Favourite struct {
+	FavTrackId    int `json:"fav_track_id"`
+	TrackId       int `json:"track_id"`
```
Comment (cp-sumi-k) [inline favourite/favourite.go]: Do we need both `FavTrackId` and `TrackId`

File: favourite/favourite.go [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+
+	err := c.ShouldBindWith(&input, binding.JSON)
+
+	if err != nil {
+		c.Abort()
+		c.JSON(http.StatusBadRequest, gin.H{"Message": err.Error()})
+		return
+	}
+
+	_, err = repository.Db.Exec(`INSERT INTO Favourite_tracks (track_id,user_id,fav_track_index) VALUES (?,?,?)`, input.TrackId, input.UserID, input.FavTrackIndex)
```
Comment (cp-sumi-k) [inline favourite/favourite.go]: Do we need to provide track_index from input only? Or we should handle it by ourselves? What is best way?

File: favourite/favourite.go [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+	if err != nil {
+		c.AbortWithStatus(http.StatusInternalServerError)
+		return
+	}
+
+	c.JSON(http.StatusOK, input)
+
+}
+
+func (repository *FavRepository) GetTrack() (input []Favourite, err error) {
```
Comment (cp-sumi-k) [inline favourite/favourite.go]: If I want my favorite track, then will I get `trackIDs` only? Is it best practice for API to return only Ids to users?

File: main.go [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+	router.PUT("/api/v1/track/edit", trackRepo.Update)
+	router.DELETE("/api/v1/track/remove", trackRepo.Delete)
+	router.GET("/api/v1/track/show", trackRepo.Read)
+
+	//Playlist
+	router.POST("/api/v1/playlist", playlistRepo.Create)
+	router.PUT("/api/v1/playlist/edit", playlistRepo.Update)
+	router.DELETE("/api/v1/playlist/remove", playlistRepo.Delete)
+	router.GET("/api/v1/playlist/show", playlistRepo.Read)
+	router.POST("/api/v1/playlist/add-track-playlist", playlistRepo.Add)
```
Comment (cp-sumi-k) [inline main.go]: I think its name should be `/api/v1/playlist/track` And method name should be `AddTrackToPlaylist`

File: main.go [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+	router.DELETE("/api/v1/playlist/remove-track-playlist", playlistRepo.Remove)
```
Comment (cp-sumi-k) [inline main.go]: Same as above comment

File: register/register.go [hunk 22]
```diff
@@ -0,0 +1,243 @@
...
+	Username string `json:"username"`
+	UserType string `json:"user_type"`
+	jwt.StandardClaims
+} //jwt service
+type JWTService interface {
+	GenerateToken(email string, isUser bool) string
+	ValidateToken(token string) (*jwt.Token, error)
+}
+
+type authCustomClaims struct {
```
Comment (cp-sumi-k) [inline register/register.go]: Why do we need two claims structures? those are `authCustomClaims` and `Claims`

File: register/register.go [hunk 23]
```diff
@@ -0,0 +1,243 @@
...
+		if _, isvalid := token.Method.(*jwt.SigningMethodHMAC); !isvalid {
+			return nil, fmt.Errorf("Invalid token", token.Header["alg"])
+
+		}
+		return []byte(register.secretKey), nil
+	})
+
+}
+
+// Create the JWT key used to create the signature
```
Comment (cp-sumi-k) [inline register/register.go]: Remove extra code

File: register/register.go [hunk 24]
```diff
@@ -0,0 +1,243 @@
...
+	input := UserRegister{}
+	err := c.ShouldBindWith(&input, binding.JSON)
+
+	if err != nil {
+		c.Abort()
+		c.JSON(http.StatusBadRequest, gin.H{"message": err.Error()})
+		return
+	}
+	auth := JWTAuthService().GenerateToken(input.Email, true)
+	_, err = repository.Db.Exec(`INSERT INTO Users(name,email,user_type,auth_token) VALUES (?,?,?,?)`, input.Name, input.Email, "General", auth)
```
Comment (cp-sumi-k) [inline register/register.go]: Is user_type `general`? How? Can't we take it as like Admin and User?

File: register/register.go [hunk 25]
```diff
@@ -0,0 +1,243 @@
...
+	if err != nil {
+		panic(err)
+	}
+
+	c.JSON(http.StatusOK, gin.H{"message": "Register Successfully"})
+
+}
+
+// ADMIN REGISTER
+func (repository *RegisterRepository) RegisterAdmin(c *gin.Context) {
```
Comment (cp-sumi-k) [inline register/register.go:122]: I think we can make common method for both registration

File: register/register.go [hunk 26]
```diff
@@ -0,0 +1,243 @@
...
+		panic(err)
+	}
+
+	c.JSON(http.StatusOK, gin.H{"message": "Admin Registered Successfully"})
+
+}
+
+// LOGIN
+
+func (repository *RegisterRepository) LoginCredential(c *gin.Context) LoginService {
```
Comment (cp-sumi-k) [inline register/register.go]: Is this method name appropriate? What is use of this method?

File: register/register.go [hunk 27]
```diff
@@ -0,0 +1,243 @@
...
+	var email, utype, token string
+
+	if err := c.ShouldBindJSON(&input); err != nil {
+		c.JSON(http.StatusUnprocessableEntity, "Invalid json provided")
+
+	}
+
+	_ = repository.Db.Get(&utype, `SELECT user_type FROM Users WHERE email= ?`, input.Email)
+	err := repository.Db.Get(&email, `SELECT email FROM Users WHERE email= ? AND auth_token IS NOT NULL`, input.Email)
+	_ = repository.Db.Get(&token, `SELECT auth_token FROM Users WHERE email= ?`, input.Email)
```
Comment (cp-sumi-k) [inline register/register.go]: Can't we make single query?

File: register/register.go [hunk 28]
```diff
@@ -0,0 +1,243 @@
...
+
+	if email != input.Email {
+		c.JSON(http.StatusUnauthorized, "Please provide valid login details")
+
+	} else if err != nil {
+		panic(err)
+
+	}
+
+	return &loginInformation{
```
Comment (cp-sumi-k) [inline register/register.go]: Shouldn't we have to return JSON response?

File: register/register.go [hunk 29]
```diff
@@ -0,0 +1,243 @@
...
+
+// LOGIN AUTHENTICATION
+type LoginService interface {
+	LoginUser(email string) bool
+}
+type loginInformation struct {
+	email string
+}
+
+func StaticLoginService() LoginService {
```
Comment (cp-sumi-k) [inline register/register.go]: Is it needed?

File: track/track.go [hunk 30]
```diff
@@ -0,0 +1,116 @@
...
+	"github.com/gin-gonic/gin"
+	"github.com/gin-gonic/gin/binding"
+	_ "github.com/go-sql-driver/mysql"
+	"github.com/jmoiron/sqlx"
+)
+
+type Track struct {
+	Id          int    `json:"track_id"`
+	Name        string `json:"name"`
+	TrackIndex  int    `json:"track_index"`
```
Comment (cp-sumi-k) [inline track/track.go:15]: Why do we need `track_index` in `tracks` table?

File: register/register.go [hunk 31]
```diff
@@ -0,0 +1,243 @@
...
+	err := ctx.ShouldBind(&credential)
+	if err != nil {
+		return "no data found"
+	}
+	isUserAuthenticated := register.loginService.LoginUser(credential.Email)
+	if isUserAuthenticated {
+		return register.jWtService.GenerateToken(credential.Email, true)
+
+	}
+	return ""
```
Comment (cp-sumi-k) [inline register/register.go]: Will we return empty string after login?

**Review by cp-sumi-k — CHANGES_REQUESTED** [review]

File: README.md [hunk 32]
```diff
@@ -1,2 +1,1036 @@
...
+- Response :
+
+    - If any server error is there then,
+    - Status Code: 500 Internal server error
+    - Data: no response data
+
+    - If request will success ,
+    - Status Code: 200 Ok
+    - Headers : none
+    - Body :{
```
Comment (cp-sumi-k) [inline README.md]: Do we need to provide a body in response or request?

File: README.md [hunk 33]
```diff
@@ -1,2 +1,1036 @@
...
+    },
+    {
+        "id": 0,
+        "name": "",
+        "image_url": "",
+        "created_at": "2022-03-25 09:38:32",
+        "updated_at": "2022-03-25 09:38:32"
+    },
+    {
+        "id": 0,
```
Comment (cp-sumi-k) [inline README.md:271]: Is id 0?

File: main.go [hunk 34]
```diff
@@ -0,0 +1,167 @@
...
+			}
+		}
+		fmt.Println(err)
+		c.AbortWithStatus(http.StatusUnauthorized)
+
+	}
+}
+
+//Middleware for user
+func UserAuthorizeJWT() gin.HandlerFunc {
```
Comment (cp-sumi-k) [inline main.go]: I think, it is not used by any routes, what is its purpose?

File: main.go [hunk 35]
```diff
@@ -0,0 +1,167 @@
...
+	return func(c *gin.Context) {
+
+		authHeader := c.GetHeader("Token")
+		tokenString := authHeader
+		token, err := jwt.Parse(tokenString, func(token *jwt.Token) (interface{}, error) {
```
Comment (cp-sumi-k) [inline main.go]: We have to check token is valid or not in one middleware, and in the other, we have to check the role of that token, 
Here work of both middleware is the same, then what is the meaning of taking two same middleware?

File: main.go [hunk 36]
```diff
@@ -0,0 +1,167 @@
...
+		authorized.DELETE("/v1/playlist/remove", playlistRepo.Delete)
+
+		authorized.POST("/v1/playlist/add-track-playlist", playlistRepo.AddPlaylist)
+		authorized.DELETE("/v1/playlist/remove-track-playlist", playlistRepo.Remove)
+
+	}
+
+	//Favourite Track (User functionality API)
+
+	authorizedUser := router.Group("/api", AuthorizeJWT())
```
Comment (cp-sumi-k) [inline main.go]: What is different between `authorized` and `authorizedUser` as both are using same middleware?

File: register/register.go [hunk 37]
```diff
@@ -0,0 +1,200 @@
...
+
+	if err != nil {
+		panic(err)
+		return
+	}
+
+	var tokenDetail Token
+	tokenDetail.Email = token.Email
+	tokenDetail.Role = token.Role
+	tokenDetail.TokenString = validToken["access_token"]
```
Comment (cp-sumi-k) [inline register/register.go:198]: Why do we need to set token detail?

**Review by cp-sumi-k — CHANGES_REQUESTED** [review]

File: README.md (hunk 33 above)
Comment (cp-priyanka-g) [reply README.md:271]: i have not fetch id data from table that's why it shows zero .

File: register/register.go (hunk 37 above)
Comment (cp-priyanka-g) [reply register/register.go:198]: i have used to check whether i was getting proper data or not ..i will remove it .

File: main.go (hunk 36 above)
Comment (cp-priyanka-g) [reply main.go]: authorized is for admin and authorizeUser for user login

**Pull request merged**